### PR TITLE
Stacksets

### PR DESCRIPTION
**Pull request by skrisyuk**
This PR introduces StackSets support to rain.

### CODE REVIEW

File: internal/aws/cfn/cfn.go [hunk 1]
```diff
@@ -126,6 +126,91 @@ func ListStacks() ([]types.StackSummary, error) {
 	return stacks, nil
 }
 
+// ListStackSets returns a list of all existing stack sets
+func ListStackSets() ([]types.StackSetSummary, error) {
+	stacks := make([]types.StackSetSummary, 0)
```
Comment (ericzbeard) [inline internal/aws/cfn/cfn.go:131]: Should this variable be called `stackSets` ?

File: internal/aws/cfn/cfn.go [hunk 2]
```diff
@@ -136,6 +221,63 @@ func DeleteStack(stackName string) error {
...
+// DeleteStackSet deletes a stack set
+func DeleteStackSet(stackSetName string) error {
+	_, err := getClient().DeleteStackSet(context.Background(), &cloudformation.DeleteStackSetInput{
+		StackSetName: &stackSetName,
+	})
+
+	return err
+}
+
+// DeleteStackSet deletes a stack set
```
Comment (ericzbeard) [inline internal/aws/cfn/cfn.go:233]: Comment doesn't match function name.

File: internal/aws/cfn/cfn.go [hunk 3]
```diff
@@ -136,6 +221,63 @@ func DeleteStack(stackName string) error {
...
+func DeleteAllChangeSetInstances(stackSetName string, wait bool) error {
```
Comment (ericzbeard) [inline internal/aws/cfn/cfn.go:234]: Should this be `DeleteAllStackSetInstances` ?

**Comment by skrisyuk** [comment]
fixed all the issue mentioned here

File: README.md [hunk 4]
```diff
@@ -21,6 +21,8 @@ Key features:
 
 * **Build new CloudFormation templates**: `rain build` generates new CloudFormation templates containing skeleton resources that you specify. This saves you having to look up which properties are available and which are required vs. optional.
 
+* **Manipulate CloudFormation stack sets**: `rain stackset deploy` creates new stackset, updates existing one or adds stack instance(s) to an existing stack set. You can list stack sets using `rain stackset ls`, review stack set details with `rain stackset ls <stack set name>` and delete stack set and\or its instances with `rain stackset rm <stack set name>`
```
Comment (ConnorKirk) [inline README.md:24]:
```suggestion
* **Manipulate CloudFormation stack sets**: `rain stackset deploy` creates a new stackset, updates an existing one or adds a stack instance(s) to an existing stack set. You can list stack sets using `rain stackset ls`, review stack set details with `rain stackset ls <stack set name>` and delete stack set and\or its instances with `rain stackset rm <stack set name>`
```

File: cft/format/json.go [hunk 5]
```diff
@@ -80,3 +80,8 @@ func convertToJSON(in string) string {
 	}
 	return string(out)
 }
+
```
Comment (ericzbeard) [inline cft/format/json.go:83]: Should have a doc comment

File: internal/aws/cfn/common.go [hunk 6]
```diff
@@ -5,7 +5,7 @@ import (
 	"github.com/aws/smithy-go/ptr"
 )
 
-func makeTags(tags map[string]string) []types.Tag {
+func MakeTags(tags map[string]string) []types.Tag {
```
Comment (ericzbeard) [inline internal/aws/cfn/common.go:8]: Doc comment

File: internal/aws/cfn/common.go [hunk 7]
```diff
@@ -17,3 +17,18 @@ func makeTags(tags map[string]string) []types.Tag {
 
 	return out
 }
+
+// uniqueStrings returns a unique subset of the string slice provided.
+func uniqueStrings(input []string) []string {
```
Comment (ericzbeard) [inline internal/aws/cfn/common.go:22]: We should have unit tests for deterministic helper functions like this.

File: internal/cmd/stackset/stackset.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+		c.Flags().StringVarP(&config.Profile, "profile", "p", "", "AWS profile name; read from the AWS CLI configuration file")
+		c.Flags().StringVarP(&config.Region, "region", "r", "", "AWS region to use")
+	}
+
+	StackSetCmd.AddCommand(c)
+}
+
+var StackSetCmd = &cobra.Command{
+	Use:   "stackset <stack_set command>",
+	Short: "This command allows to manipulate stack sets.",
```
Comment (ericzbeard) [inline internal/cmd/stackset/stackset.go:44]:
```suggestion
	Short: "This command manipulates stack sets.",
```

File: internal/cmd/stackset/stackset.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	Long:  "This command allows to manipulate stack sets. It has no action if specific stack set command is not added.",
```
Comment (ericzbeard) [inline internal/cmd/stackset/stackset.go:45]:
```suggestion
	Long:  "This command manipulates stack sets. It has no action if specific stack set command is not added.",
```

File: internal/cmd/stackset/stackset.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+func init() {
+	addCommand(true, StackSetLsCmd)
+	addCommand(true, StackSetDeployCmd)
+	addCommand(true, StackSetRmCmd)
+
+	oldUsageFunc := StackSetCmd.UsageFunc()
```
Comment (ericzbeard) [inline internal/cmd/stackset/stackset.go:53]: I'm a little confused about what's going on here, can you add a comment?

File: internal/cmd/stackset/stackset_deploy.go [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+	var existingRegions []string
+
+	for _, instance := range instances {
+		existingRegions = append(existingRegions, *instance.Region)
+	}
+	c.Regions = difference(c.Regions, existingRegions)
+}
+
+// difference returns the elements in `a` that aren't in `b`.
+func difference(a, b []string) []string {
```
Comment (ericzbeard) [inline internal/cmd/stackset/stackset_deploy.go:279]: Add unit tests for this function and others like it.

File: internal/cmd/stackset/stackset_deploy.go [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+
+		// Create Stack Set instances
+		spinner.Push("Creating stack set instances")
+		err = cfn.CreateStackSetInstances(stackSetInstancesConfig, !detach)
+		spinner.Pop()
+		if err != nil {
+			panic(ui.Errorf(err, "error while creating stack set instances"))
+		}
+		if !detach {
+			fmt.Println("Stack set instances have been created successfuly")
```
Comment (ericzbeard) [inline internal/cmd/stackset/stackset_deploy.go:351]:
```suggestion
			fmt.Println("Stack set instances have been created successfully")
```

File: internal/aws/cfn/cfn_types.go [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+	// service fields, not to be used in configuration file
+	StackSetName string            `yaml:"-"`
+	Template     cft.Template      `yaml:"-"`
+	Parameters   []types.Parameter `yaml:"-"`
+	Tags         []types.Tag       `yaml:"-"`
+}
+
+// we treat list of provided accounts and regions as requirement to have instances in all the provided
+// accounts weather updated or created(added).
```
Comment (ericzbeard) [inline internal/aws/cfn/cfn_types.go:158]:
```suggestion
// accounts whether updated or created(added).
```

File: internal/cmd/stackset/stackset_deploy.go [hunk 14]
```diff
@@ -0,0 +1,426 @@
...
+	}
+
+	c.Accounts = intersection(c.Accounts, existingAccounts)
+	c.Regions = intersection(c.Regions, existingRegions)
+}
+
+// removes existing instances from the StackSetInstancesConfig.
+// We do not remove accounts because we accept list of provided
+// accounts and regions as requirement to have instances in all provided
+// accounts weather updated or created(added)
```
Comment (ericzbeard) [inline internal/cmd/stackset/stackset_deploy.go:262]:
```suggestion
// accounts whether updated or created(added)
```

**Review by ericzbeard — CHANGES_REQUESTED** [review]
Overall this is great work! Just a few minor changes and requests for some unit tests on deterministic functions.

**Comment by ericzbeard** [comment]
Also looks like we have some linter checks failing, let's address those.

File: internal/cmd/stackset/stackset.go (hunk 10 above)
Comment (skrisyuk) [reply internal/cmd/stackset/stackset.go:53]: Here we add sub commands to a stackset command

File: internal/aws/cfn/common.go (hunk 7 above)
Comment (skrisyuk) [reply internal/aws/cfn/common.go:22]: +1

**Review by ericzbeard — APPROVED** [review]

**Pull request merged**